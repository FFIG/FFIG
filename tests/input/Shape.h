#include <cmath>

// Any class annotated with this macro will be exposed and have all functions
// exposed.
// Any class deriving from an exposed class will have its constructors exposed.
#ifdef __clang__
#define C_API __attribute__((annotate("GENERATE_C_API")))
#else
#define C_API
#endif

struct AbstractShape
{
  virtual ~AbstractShape()
  {
  }
  virtual double area() const = 0;
  virtual double perimeter() const = 0;
  virtual const char* name() const = 0;
  virtual int is_equal(const AbstractShape* s) const = 0;
} C_API;

static const double pi = 3.14159265359;

class Circle : public AbstractShape
{
  const double radius_;

public:
  double area() const override
  {
    return pi * radius_ * radius_;
  }

  double perimeter() const override
  {
    return 2 * pi * radius_;
  }

  const char* name() const override
  {
    return "Circle";
  }

  int is_equal(const AbstractShape* s) const override
  {
    if ( auto c = dynamic_cast<const Circle*>(s) )
      return c->radius_ == radius_;
    return false;
  }

  Circle(double radius) : radius_(radius)
  {
    if ( radius < 0 ) 
    { 
      std::string s = "Circle radius \"" + std::to_string(radius_) + "\" must be non-negative.";
      throw std::runtime_error(s);
    }
  }
};

class Square : public AbstractShape
{
  const double side_;

public:
  double area() const override
  {
    return side_ * side_;
  }

  double perimeter() const override
  {
    return 4.0 * side_;
  }

  const char* name() const override
  {
    return "Square";
  }

  int is_equal(const AbstractShape* s) const override
  {
    if ( auto sq = dynamic_cast<const Square*>(s) )
      return sq->side_ == side_;
    return false;
  }

  Square(double side) : side_(side)
  {
    if ( side_ <= 0.0 )
    {
      std::string s = "Square side \"" + std::to_string(side_) + "\" must be non-negative.";
      throw std::runtime_error(s);
    }
  }
};

class Pentagon : public AbstractShape
{
  const double side_;

public:
  double area() const override
  {
    return 0.25 * sqrt(5. * (5. + 2. * sqrt(5.))) * side_ * side_;
  }

  double perimeter() const override
  {
    return 5.0 * side_;
  }

  const char* name() const override
  {
    return "Pentagon";
  }

  int is_equal(const AbstractShape* s) const override
  {
    if ( auto p = dynamic_cast<const Pentagon*>(s) )
      return p->side_ == side_;
    return false;
  }

  Pentagon(double side) : side_(side)
  {
    if ( side_ <= 0.0 )
    {
      std::string s = "Pentagon side \"" + std::to_string(side_) + "\" must be non-negative.";
      throw std::runtime_error(s);
    }
  }
};
