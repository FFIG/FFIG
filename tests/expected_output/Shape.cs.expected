// This code was generated by FFIG <http://ffig.org>.
// Manual edits will be lost.
using System;
using System.Runtime.InteropServices;

namespace Shape_c {

  public class Exception : System.Exception {

    [DllImport("Shape_c")]
    private static extern void Shape_clear_error();

    [DllImport("Shape_c")]
    private static extern IntPtr Shape_error();

    public Exception() : base(Marshal.PtrToStringAnsi(Shape_error())) {
      Shape_clear_error();
    }
  }

  public class AbstractShape {

    [DllImport("Shape_c")]
    private static extern int Shape_AbstractShape_dispose(IntPtr c_obj);
    [DllImport("Shape_c")]
    private static extern double Shape_AbstractShape_area_noexcept(IntPtr c_obj);
    [DllImport("Shape_c")]
    private static extern double Shape_AbstractShape_perimeter_noexcept(IntPtr c_obj);
    [DllImport("Shape_c")]
    private static extern IntPtr Shape_AbstractShape_name_noexcept(IntPtr c_obj);
    [DllImport("Shape_c")]
    private static extern int Shape_AbstractShape_is_equal_noexcept(IntPtr c_obj, IntPtr s);

    protected IntPtr c_obj_;

    protected AbstractShape() {
    }

    protected AbstractShape(IntPtr c_obj) {
      c_obj_ = c_obj;
    }

    ~AbstractShape() {
      Shape_AbstractShape_dispose(c_obj_);
    }
    public double area {
      get {
        var rv = Shape_AbstractShape_area_noexcept(c_obj_);
        return rv;
      }
    }
    public double perimeter {
      get {
        var rv = Shape_AbstractShape_perimeter_noexcept(c_obj_);
        return rv;
      }
    }
    public string name {
      get {
        var rv = Shape_AbstractShape_name_noexcept(c_obj_);
        return Marshal.PtrToStringAnsi(rv);
      }
    }
    public int is_equal(AbstractShape s) {
      var rv = Shape_AbstractShape_is_equal_noexcept(c_obj_, s.c_obj_);
      return rv;
    }
  }

  public class Circle : Shape_c.AbstractShape {

    [DllImport("Shape_c")]
    private static extern int Shape_Circle_create(double radius, out IntPtr ptr);

    public Circle(double radius) {
      int rc = Shape_Circle_create(radius, out c_obj_);
      if(rc != 0) {
        throw new Shape_c.Exception();
      }
    }
  }

  public class Square : Shape_c.AbstractShape {

    [DllImport("Shape_c")]
    private static extern int Shape_Square_create(double side, out IntPtr ptr);

    public Square(double side) {
      int rc = Shape_Square_create(side, out c_obj_);
      if(rc != 0) {
        throw new Shape_c.Exception();
      }
    }
  }

  public class Pentagon : Shape_c.AbstractShape {

    [DllImport("Shape_c")]
    private static extern int Shape_Pentagon_create(double side, out IntPtr ptr);

    public Pentagon(double side) {
      int rc = Shape_Pentagon_create(side, out c_obj_);
      if(rc != 0) {
        throw new Shape_c.Exception();
      }
    }
  }

}
